src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcRar.cs

```diff
@@ -0,0 +1,138 @@
+                                                            is64Bit: EnvironmentUtilities.Is64BitProcess,
+                                                            workerNode: false));
+            NodeContext context = GetNode(null, commandLineArgs, nodeId, factory, hostHandshake, NodeContextTerminated);
+
+            if (null != context)
+            {
+                _rarNodeId = nodeId;
+                _rarNodeContext = context;
+
+                // Is this necesarry?
+                // Start the asynchronous read.
+                context.BeginAsyncPacketRead();
+
+                // Configure the node.
+                context.SendData(configuration);
+
+                return true;
+            }
+
+            throw new BuildAbortedException(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("CouldNotConnectToMSBuildExe", ComponentHost.BuildParameters.NodeExeLocation));
+        }
+
+        private void NodeContextTerminated(int nodeId)
+        {
+            if (_rarNodeId == nodeId)
+            {
+                _rarNodeId = null;
+                _rarNodeContext = null;
+            }
+        }
+
+        #region IBuildComponent Members
+
+        /// <summary>
+        /// Initializes the component.
+        /// </summary>
+        /// <param name="host">The component host.</param>
+        public void InitializeComponent(IBuildComponentHost host)
+        {
+            this.ComponentHost = host;
+        }
+
+        /// <summary>
+        /// Shuts down the component
+        /// </summary>
+        public void ShutdownComponent()
+        {
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Static factory for component creation.
+        /// </summary>
+        static internal IBuildComponent CreateComponent(BuildComponentType componentType)
+        {
+            ErrorUtilities.VerifyThrow(componentType == BuildComponentType.OutOfProcNodeRarProvider, "Factory cannot create components of type {0}", componentType);
+            return new NodeProviderOutOfProcRar();
+        }
+
+        public void SendData(int node, INodePacket packet)
+        {
+            if (_rarNodeId == node)
+            {
+                _rarNodeContext.SendData(packet);
+            }
+        }
+
+        public void ShutdownAllNodes()
+        {
+            // If no BuildParameters were specified for this build,
+            // we must be trying to shut down idle nodes from some
+            // other, completed build. If they're still around,
+            // they must have been started with node reuse.
+            bool nodeReuse = ComponentHost.BuildParameters?.EnableNodeReuse ?? true;
+
+            // To avoid issues with mismatched priorities not shutting
+            // down all the nodes on exit, we will attempt to shutdown
+            // all matching nodes with and without the priority bit set.
+            // This means we need both versions of the handshake.
+
+            // RAR node is special...
```

Comment:
Why is the RAR node special?

Reply:
My thought here was that I don't want to shutdown RAR node if the node will be later reused by next build.

Reply:
That's certainly true at the end of a normal build, but if the user calls ShutdownAllNodes (and this _is_ something the user gets to call on its own), we want all nodes, including RAR, to shut down.